File: tests/test_device.py
import struct

import pytest

from cougar_rgb import device
from cougar_rgb.device import EFFECT_ZONE, Fusion2


@pytest.fixture
def dev(monkeypatch):
    sent = []

    def ioctl(fd, request, buf):
        assert request == device.HIDIOCSFEATURE
        assert len(buf) == device.PACKET_SIZE
        sent.append(bytes(buf))

    monkeypatch.setattr(device.fcntl, 'ioctl', ioctl)
    d = Fusion2.__new__(Fusion2)
    d.fd, d.path, d.sent = -1, '/dev/hidraw-test', sent
    return d


def test_command_is_padded_feature_report(dev):
    dev.command(device.CMD_APPLY, 0xFF)
    assert dev.sent == [bytes([0xCC, 0x28, 0xFF]).ljust(64, b'\0')]


@pytest.mark.parametrize('enabled, value', [(True, device.DIRECT_BIT), (False, 0)])
def test_set_direct(dev, enabled, value):
    dev.set_direct(enabled)
    assert dev.sent[0][:3] == bytes([0xCC, 0x32, value])


def test_effect_packet_layout(dev):
    dev.set_effect(EFFECT_ZONE, 2, color=(255, 16, 1), max_brightness=200, min_brightness=3,
                   periods=(1000, 900, 200, 0), params=(7, 1, 2, 0))
    buf = dev.sent[0]
    assert buf[1] == 0x20 + EFFECT_ZONE
    assert struct.unpack_from('<II', buf, 2) == (1 << EFFECT_ZONE, 0)
    assert buf[11:14] == bytes([2, 200, 3])
    assert buf[14:18] == bytes([16, 255, 1, 0])       # GRB + выравнивание
    assert buf[18:22] == bytes(4)
    assert struct.unpack_from('<4H', buf, 22) == (1000, 900, 200, 0)
    assert buf[30:34] == bytes([7, 1, 2, 0])


def test_write_leds_splits_into_19_led_packets(dev):
    colors = [(i, 100 + i, 200) for i in range(25)]
    dev.write_leds(colors)
    assert len(dev.sent) == 2
    first, second = dev.sent
    assert first[1] == device.DIRECT_REG
    assert struct.unpack_from('<HB', first, 2) == (0, 57)
    assert struct.unpack_from('<HB', second, 2) == (57, 18)
    assert first[5:8] == bytes([100, 0, 200])            # GRB
    assert second[5:8] == bytes([119, 19, 200])


def test_find_hidraw_matches_vendor_and_product(tmp_path, monkeypatch):
    for name, hid_id in (('hidraw0', '0003:000009DA:00009090'), ('hidraw2', '0003:0000048D:00005702')):
        (tmp_path / name / 'device').mkdir(parents=True)
        (tmp_path / name / 'device' / 'uevent').write_text('DRIVER=hid-generic\nHID_ID=%s\n' % hid_id)
    monkeypatch.setattr(device.glob, 'glob', lambda pattern: sorted(str(p) for p in tmp_path.iterdir()))
    assert device.find_hidraw() == '/dev/hidraw2'


def test_find_hidraw_raises_when_missing(monkeypatch):
    monkeypatch.setattr(device.glob, 'glob', lambda pattern: [])
    with pytest.raises(device.DeviceNotFound):
        device.find_hidraw()

File: tests/test_device.cpp
#include "device.hpp"
#include "fakes.hpp"

#include <catch2/catch_test_macros.hpp>
#include <filesystem>
#include <fstream>

using namespace cougar;
using test::RecordingTransport;

namespace {

struct Setup {
    RecordingTransport *transport;
    Fusion2 dev;
    Setup() : Setup(std::make_unique<RecordingTransport>()) {}
    explicit Setup(std::unique_ptr<RecordingTransport> t) : transport(t.get()), dev(std::move(t)) {}
};

uint16_t u16(const Packet &p, size_t at) { return p[at] | p[at + 1] << 8; }
uint32_t u32(const Packet &p, size_t at) { return p[at] | p[at + 1] << 8 | p[at + 2] << 16 | uint32_t(p[at + 3]) << 24; }

}  // namespace

TEST_CASE("command is a zero-padded feature report")
{
    Setup s;
    s.dev.command({CMD_APPLY, 0xFF});
    Packet expected{};
    expected[0] = 0xCC, expected[1] = 0x28, expected[2] = 0xFF;
    REQUIRE(s.transport->sent == std::vector{expected});
}

TEST_CASE("set_direct toggles the D_LED1 bit")
{
    Setup s;
    s.dev.set_direct(true);
    s.dev.set_direct(false);
    REQUIRE(s.transport->sent[0][1] == 0x32);
    REQUIRE(s.transport->sent[0][2] == DIRECT_BIT);
    REQUIRE(s.transport->sent[1][2] == 0);
}

TEST_CASE("effect packet layout")
{
    Setup s;
    s.dev.set_effect(EFFECT_ZONE, {.type = 2, .color = {255, 16, 1}, .max_brightness = 200, .min_brightness = 3,
                                   .periods = {1000, 900, 200, 0}, .params = {7, 1, 2, 0}});
    const Packet &buf = s.transport->sent.at(0);
    CHECK(buf[1] == 0x20 + EFFECT_ZONE);
    CHECK(u32(buf, 2) == 1u << EFFECT_ZONE);
    CHECK(u32(buf, 6) == 0);
    CHECK((buf[11] == 2 && buf[12] == 200 && buf[13] == 3));
    CHECK((buf[14] == 16 && buf[15] == 255 && buf[16] == 1 && buf[17] == 0));   // GRB + padding
    CHECK(u32(buf, 18) == 0);
    CHECK((u16(buf, 22) == 1000 && u16(buf, 24) == 900 && u16(buf, 26) == 200 && u16(buf, 28) == 0));
    CHECK((buf[30] == 7 && buf[31] == 1 && buf[32] == 2 && buf[33] == 0));
}

TEST_CASE("write_leds splits into 19-LED packets in GRB order")
{
    Setup s;
    std::vector<Rgb8> colors;
    for (int i = 0; i < 25; ++i)
        colors.push_back({uint8_t(i), uint8_t(100 + i), 200});
    s.dev.write_leds(colors);
    REQUIRE(s.transport->sent.size() == 2);
    const auto &first = s.transport->sent[0], &second = s.transport->sent[1];
    CHECK(first[1] == DIRECT_REG);
    CHECK((u16(first, 2) == 0 && first[4] == 57));
    CHECK((u16(second, 2) == 57 && second[4] == 18));
    CHECK((first[5] == 100 && first[6] == 0 && first[7] == 200));
    CHECK((second[5] == 119 && second[6] == 19 && second[7] == 200));
}

TEST_CASE("info parses the 0x60 report")
{
    auto transport = std::make_unique<RecordingTransport>();
    Packet &r = transport->reply;
    r[0] = 0xCC, r[1] = 1, r[4] = 3, r[5] = 5, r[6] = 5, r[7] = 0;
    const std::string name = "IT5701-GIGABYTE V3.5.5.0";
    std::copy(name.begin(), name.end(), r.begin() + 12);
    r[56] = 0x00, r[57] = 0x01, r[58] = 0x01, r[59] = 0x57;
    Setup s(std::move(transport));
    const DeviceInfo info = s.dev.info();
    CHECK(s.transport->sent.at(0)[1] == CMD_INFO);
    CHECK(info.firmware == "3.5.5.0");
    CHECK(info.name == name);
    CHECK(info.chip_id == "0x57010100");
}

TEST_CASE("find_hidraw matches vendor and product")
{
    const auto root = std::filesystem::temp_directory_path() / "cougar-rgb-test-sysfs";
    std::filesystem::remove_all(root);
    for (auto [name, id] : {std::pair{"hidraw0", "0003:000009DA:00009090"}, {"hidraw2", "0003:0000048D:00005702"}}) {
        std::filesystem::create_directories(root / name / "device");
        std::ofstream(root / name / "device" / "uevent") << "DRIVER=hid-generic\nHID_ID=" << id << "\n";
    }
    CHECK(find_hidraw(root) == "/dev/hidraw2");
    std::filesystem::remove_all(root / "hidraw2");
    CHECK_THROWS_AS(find_hidraw(root), DeviceNotFound);
    std::filesystem::remove_all(root);
}
